src/build/vite/plugin.ts

```diff
@@ -238,6 +239,44 @@ function nitroMain(ctx: NitroPluginContext): VitePlugin {
       debug("[main] Configuring dev server");
       return configureViteDevServer(ctx, server);
     },
+
+    // automatically reload the client when a server module is updated
+    // see: https://github.com/vitejs/vite/issues/19114
+    hotUpdate(options) {
+      if (
+        !ctx.pluginConfig.config?.devServer?.serverReload ||
+        this.environment.name === "client"
+      )
+        return;
+
+      let hasServerOnlyModule = false;
+
+      const invalidated = new Set<EnvironmentModuleNode>();
+
+      for (const mod of options.modules) {
+        if (!mod.id) continue;
+
+        // check if module exists in the client module graph
+        const clientModule =
+          options.server.environments.client.moduleGraph.getModuleById(mod.id);
+        // if so, the client env will handle the update
+        if (clientModule) continue;
+
+        // must be a module that is only SSR, invalidate it
+        this.environment.moduleGraph.invalidateModule(
+          mod,
+          invalidated,
+          options.timestamp,
+          false
+        );
+
+        hasServerOnlyModule = true;
+      }
+
+      if (hasServerOnlyModule) {
+        options.server.ws.send({ type: "full-reload" });
```

Comment:
Looking back https://v6.vite.dev/guide/migration.html, it has `return []` which is probably for skipping Vite's own invalidation handling since it's handled above already.

```suggestion
        options.server.ws.send({ type: "full-reload" });
        return [];
```